Automattic/wp-calypso
# Plans 2023: Decouple CTA actions from plans-grid-next package ( Attempt 2 )

Related to https://github.com/Automattic/wp-calypso/pull/86886

## Proposed Changes

Description taken from https://github.com/Automattic/wp-calypso/pull/86886 and amended where appropriate for this second attempt.

This PR is the first step to fully decouple the CTA configuration of the plans grid, inspired by @jdc91 's post pdvytD-qk-p2. Conceptually, it's an extension to the current `PlanActionOverrides` system, but it only replaces the callback functions part for now as the 1st step. 
The ultimate goal is that the `plans-grid-next` package simply offers the `PlanActions` prop for the calling site to configure the clicked actions, text, disabled or not, etc. The package itself shouldn't need to know those contextual info.

Summarizing the introduced changes:

* Create a `useActionCallback` hook that encapsulates all logic that determines what CTA function to use in a given circumstance
* Note that the upgrade handler can obviously be simplified, but the current structure is generally maintained for now so as to avoid introducing too much change at once.
* Introduce an action handler mechanism, i.e. `showModalAndExit`, which allows the calling site to introduce further actions on the CTA click. It also allows returning a boolean value to indicate whether it wants to trigger early return, so the actual shopping cart processing won't happen. Two cases supported by this are the domain/plan upsell modal and the domain upsell from home modal.
* Deprecate the current CTA callback props, e.g. onUpgradeClick, handleUpgradeClick, `callback` in `PlanActionOverrides`, etc.
* The actual on-click callback has been separated finer, e.g. managing add-ons, managing plans, redirecting to the VIP page, handling upgrades, instead of being combined into one like it currently does. 

## Testing Instructions



**Onboarding Ctas**
- Navigate to `/start/plans`
- Smoke test plan ctas in the features grid ( free, explorer, creator, etc. ) and ensure that the correct plan is added to the checkout cart
- Smoke test creator or entrepreneur plan ctas with a storage add-on and ensure the correct products are added to the checkout cart
- Repeat for the comparison grid
- Lists of flows with the plans pricing page can be found here PCYsg-RDT-p2
- Repeat for a different signup flow ( Ex. `start/domain` ) 
- Repeat for a stepper flow ( Ex. `setup/newsletter` ) 

**Paid Domain Modal Upsell Ctas**
- Navigate to `/start/onboarding-pm`
- Select a paid, custom domain
- Select a free plan
- Verify that the upsell modal is displayed as expected
- Smoke test the ctas in the upsell modal

<img width="725" alt="Screenshot 2024-04-23 at 4 04 39 PM" src="https://github.com/Automattic/wp-calypso/assets/5414230/537ca848-ac0a-4c99-b1fc-d5c4b08dc59c">

**Free Domain Free Plan Modal Upsell Ctas**
- Navigate to `/start/onboarding-pm`
- Select a free domain
- Select a free plan
- Verify that the upsell modal is displayed as expected
- Smoke test the ctas in the upsell modal

<img width="664" alt="Screenshot 2024-04-23 at 4 04 10 PM" src="https://github.com/Automattic/wp-calypso/assets/5414230/dc8af121-2e1a-401d-85f6-0e72701faa79">

**Plans Page Ctas**
- Navigate to `/plans` for a specific site Ex. `/plans/testsite123.wordpress.com`
- Smoke test plan ctas ( free, explorer, creator, etc. ) and ensure that the correct plan is added to the checkout cart
- Smoke test creator or entrepreneur plan ctas with a storage add-on and ensure the correct products are added to the checkout cart
- Repeat for the comparison grid
- Smoke test ctas for the spotlight plan ( plan card rendered at the top of the page )
  - Ensure the manage add-ons button redirects to the `/add-ons` page
  - Ensure the manage plan button redirects to the plan management page

**Test Free Trials Plan Ctas**
- Navigate to `/setup/new-hosted-site`
- Verify that the Creator plan free trial CTA behaves as it does in production ( leads to the `/trialAcknowledge` page )
- Verify that a free trial site is created when continuing through onboarding
- Smoke test the other Creator and Entrepreneur plan CTA buttons

<img width="824" alt="Screenshot 2024-04-23 at 4 12 31 PM" src="https://github.com/Automattic/wp-calypso/assets/5414230/560b9521-3fb4-4e2d-88f5-a353290df9e8">

## Pre-merge Checklist
